### PR TITLE
Add GLM4 support to PaddleFormers

### PR DESCRIPTION
**Pull request by KissrainKlein**
添加GLM4模型文件到 PaddleFormers/paddleformers/transformers/glm4
测试test_glm4.py到PaddleFormers/tests/transformers/glm4
1. 单卡前向精度对齐验证
日志显示，你的模型在 Paddle 环境下与 Hugging Face 环境下的前向 Logits 差异极小。
最大绝对误差 (max_abs_diff) 均在 1e-5 量级（例如 CASE 1 为 7.34e-05，CASE 4 为 3.33e-05）。
这远远优于你要求的 1e-2 量级，说明模型权重加载和前向计算的计算逻辑（包括 RMSNorm、RotaryEmbedding、Attention 等算子的转换）是完全正确且高精度对齐的。
2. 模型生成验证
通过对比 dump_hf_generate.py 和 dump_paddle_generate.py 的输出，两端的前 10 个生成 Token 完美一致：
CASE 1: [109377, 3837, 122786, 104668, 113255, 3837, 118295, 107410, 98734, 1773] ✅
CASE 2: [99334, 98316, 101881, 106552, 3837, 100039, 111661, 102733, 98327, 102209] ✅
CASE 3: [103974, 105678, 99917, 100000, 3837, 103153, 100370, 106889, 98366, 3837] ✅
CASE 4: [12089, 13, 1084, 374, 825, 315, 279, 1429, 6233, 9716] ✅
CASE 5: [785, 9101, 374, 1602, 1661, 3351, 13, 118291, 98516, 98500] ✅
不仅前 10 个 Token 一致，日志显示的 32 个长度的完整 generated_text 内容也是一模一样的。

### CODE REVIEW

**Comment by a31413510** [comment]
1. glm4 未注册到 Auto 映射
  PR 添加了 glm4/ 模块但未注册到 CONFIG_MAPPING_NAMES、MAPPING_NAMES、__init__.py，导致 AutoConfig 和 AutoModelForCausalLM 找不到模型类。
修改意见：手动添加到 auto/configuration.py、auto/modeling.py 和 transformers/__init__.py

2. bf16 推理时 dtype 不一致
  apply_rotary_pos_emb 中 cos/sin 是 float32，而 query/key 是 bfloat16，concat 时类型不一致报错。
修改意见：添加 .cast(q.dtype) 转换

3. 训练 loss 双移位问题
  glm4/modeling.py 中有 HF 风格的 shift_labels = labels[:, 1:]，而 PF 的 SFTDataset 已经预移位了 labels，导致双重移位。
修改意见：去掉 forward 中的 shift，直接用 logits vs labels 计算 CE loss

4. load_checkpoint_format 问题
  flex_checkpoint 需要模型实现 _gen_aoa_config，GLM-4 没有。
修改意见：添加 _gen_aoa_config 实现

**Comment by a31413510** [comment]
需要解决代码冲突

**Comment by a31413510** [comment]
需要解决代码冲突
